### PR TITLE
Create a RunningAzureFunctions class to handle multiple function tasks in one workspace

### PR DESCRIPTION
**Pull request by nturinski**
Previously, there was a limitation of only being able to run one task per workspace. This would make it impossible to run and manage multiple tasks in subfolders which was blocking a scenario for the C# Dev Kit team.

These changes make the workspace folder a key to the map, but we are storing it in an array. The array contains the cwd of the subpath for the project. This will allow us to run multiple projects in different folders simultaneously which should hopefully get us one step closer to monorepo support.

### CODE REVIEW

File: src/funcCoreTools/funcHostTask.ts [hunk 1]
```diff
@@ -79,34 +68,57 @@ export function registerFuncHostTaskEvents(): void {
         // Used to stop the task started with pickFuncProcess.ts startFuncProcessFromApi.
         if (DotnetDebugDebugConfiguration.is(debugSession.configuration) && debugSession.configuration.launchServiceData.buildPath) {
             const buildPathUri: vscode.Uri = vscode.Uri.file(debugSession.configuration.launchServiceData.buildPath)
-            stopFuncTaskIfRunning(buildPathUri.fsPath, /* terminate */ true)
+
+            const workspaceFolder = buildPathToWorkspaceFolderMap.get(debugSession.configuration.launchServiceData.buildPath)
+            if (workspaceFolder === undefined) {
+                throw Error()
```
Comment (WardenGnaw) [inline src/funcCoreTools/funcHostTask.ts:74]: nit: I wasn't sure if throwing here is correct.

Maybe a `vscode.window.showWarningMessage('Unable to find workspaceFolder for debugConfiguration.buildPath");` + however its localized

File: src/commands/pickFuncProcess.ts [hunk 2]
```diff
@@ -32,32 +33,37 @@ export async function startFuncProcessFromApi(
...
+                }
+            }
+
+            await waitForPrevFuncTaskToStop(workspaceFolder);
+
+            buildPathToWorkspaceFolderMap.set(buildPath, workspaceFolder);
+
+            const funcTask = new vscode.Task({ type: `func  ${buildPath}` },
+                workspaceFolder,
+                hostStartTaskName, `func`,
```
Comment (WardenGnaw) [inline src/commands/pickFuncProcess.ts:59]: nit: place on next line since its another argument

File: src/commands/startFunctionApp.ts [hunk 3]
```diff
@@ -5,16 +5,20 @@
...
+import { startFuncProcessFromApi } from './pickFuncProcess';
 
 export async function startFunctionApp(context: IActionContext, node?: SlotTreeItem): Promise<void> {
     if (!node) {
         node = await pickFunctionApp(context);
     }
 
+
     const client: SiteClient = await node.site.createClient(context);
+    await startFuncProcessFromApi(`${workspace.workspaceFolders![0].uri.path}/bin/Debug/net8.0`, [], {});
```
Comment (WardenGnaw) [inline src/commands/startFunctionApp.ts:21]: remove?

File: src/commands/pickFuncProcess.ts [hunk 4]
```diff
@@ -32,32 +33,37 @@ export async function startFuncProcessFromApi(
...
-            const funcTask = new vscode.Task(azFuncTaskDefinition,
-                vscode.TaskScope.Global,
-                hostStartTaskName, 'func',
+            let workspaceFolder: vscode.WorkspaceFolder | undefined = buildPathToWorkspaceFolderMap.get(buildPath);
+
+            if (workspaceFolder === undefined) {
+                workspaceFolder = {
+                    uri: vscode.Uri.parse(buildPath),
+                    name: buildPath,
+                    index: -1
```
Comment (WardenGnaw) [inline src/commands/pickFuncProcess.ts:49]: I'm also not sure if the consequences of using index: -1 if VS Code actually looks at this 😓

File: src/commands/startFunctionApp.ts (hunk 3 above)
Comment (nturinski) [reply src/commands/startFunctionApp.ts:21]: Yep, will do. Was from testing 😅

**Review by MicroFish91 — APPROVED** [review]
Approving for now so as not to block.  Ping if you need review again later

**Review by MicroFish91 — APPROVED** [review]

**Pull request merged**